### PR TITLE
Change lineage list response into a lazy-fetching generator

### PR DESCRIPTION
**Pull request by cmgrote**

### CODE REVIEW

File: pyatlan/client/atlan.py [hunk 1]
```diff
@@ -287,6 +282,56 @@ def __iter__(self) -> Generator[Asset, None, None]:
...
+            assets: list[Asset],
+        ):
+            self._client = client
+            self._criteria = criteria
+            self._start = start
+            self._size = size
+            self._has_more = has_more
+            self._assets = assets
+
+        def has_more(self) -> bool:
```
Comment (ErnestoLoma) [inline pyatlan/client/atlan.py:302]: Add the @property annotation to make this a property

File: pyatlan/client/atlan.py [hunk 2]
```diff
@@ -287,6 +282,56 @@ def __iter__(self) -> Generator[Asset, None, None]:
                 if not self.next_page():
                     break
 
+    class LineageResults:
```
Comment (ErnestoLoma) [inline pyatlan/client/atlan.py:285]: There's a lot of duplicate code between LineageResults and SearchResults. So I think maybe the following refactoring would be in good:

1. Create an abstract class SearchRequest which both LineageListRequest and IndexSearchRequest extend.
2. Change SearchResults to take a criteria that is a SearchRequest
3. Create abstract properties in SearchRequest for:
    1. has_more
    2. count
    4. offset
    5. size

**Review by ErnestoLoma — CHANGES_REQUESTED** [review]
I suggested a bit of refactoring

File: pyatlan/client/atlan.py (hunk 2 above)
Comment (cmgrote) [reply pyatlan/client/atlan.py:285]: I think only `offset` and `size` are shared:
- Only lineage list has the `has_more` (does not have a `count`)
- Only a indexsearch has a `count` (but does not have a `has_more`)

Comment (ErnestoLoma) [reply pyatlan/client/atlan.py:285]: That's why I suggest added different implementations to the sub classes that way we can eliminate the duplication.

**Review by ErnestoLoma — APPROVED** [review]

**Pull request merged**